### ISSUE TITLE
MediatR pre/post/exception handlers

### ISSUE THREAD

**Issue by martinothamar** [martinothamar/Mediator#6]
Currently don't have these. Not sure I want to add these "by default", as they will slow down all pipelines even if not used, but maybe as a separate package? Needs investigation

**Comment by ercinakyuz** [comment]
I need this too, also behaviors and pre/post processors should be also used for NotificationHandlers.
You can add it into PipelineBehavior default implementation decorate next method like this:

```
    public class PipelineBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
        where TRequest : IRequest<TResponse>
    {
        private readonly ISet<IPreProcessor<TRequest>> _preProcessors;

        private readonly ISet<IPostProcessor<TRequest,TResponse>> _postProcessors;

        public PipelineBehavior(ISet<IPreProcessor<TRequest>> preProcessors,
            ISet<IPostProcessor<TRequest, TResponse>> postProcessors)
        {
            _preProcessors = preProcessors;
            _postProcessors = postProcessors;
        }

        public async Task<TResponse> Handle(TRequest request, RequestHandlerDelegate<TResponse> next,
            CancellationToken cancellationToken)
        {
            await ExecutePreProcessorsAsync(request, cancellationToken);
            var response = await next();
            await ExecutePostProcessorsAsync(request, response, cancellationToken);
            return response;
        }

        private async Task ExecutePreProcessorsAsync(TRequest request, CancellationToken cancellationToken)
        {
            foreach (var preProcessor in _preProcessors)
            {
                await preProcessor.ProcessAsync(request, cancellationToken);
            }
        }

        private async Task ExecutePostProcessorsAsync(TRequest request, TResponse response,
            CancellationToken cancellationToken)
        {
            foreach (var postProcessor in _postProcessors)
            {
                await postProcessor.ExecuteAsync(request, response, cancellationToken);
            }
        }
    }
```

**Comment by martinothamar** [comment]
Started work on this in #67. This can be included in version 2.1. It's open for review if you are interested. As mentioned in the PR, I avoided the default implementation of the pipeline behavior as it introduces overhead even for users that don't use any pre/prost processors or exception handlers

**Comment by martinothamar** [comment]
This has been included in release [2.1.0-preview.8](https://www.nuget.org/packages/Mediator.SourceGenerator/2.1.0-preview.8)
